File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.py
import heapq
def solution(n, edge):
    answer = 1
    graph=[[] for _ in range(n+1)]
    for s,e in edge:
        graph[s].append((e,1))
        graph[e].append((s,1))

    def dijkstra(start):
        INF=int(1e9)
        distance=[INF]*(n+1)
        hq=[]
        heapq.heappush(hq,(0,start))
        distance[start]=0
        while hq:
            dist,now=heapq.heappop(hq)
            if distance[now]<dist:
                continue
            for nxt,d in graph[now]:
                cost=d+dist
                if distance[nxt]>cost:
                    distance[nxt]=cost
                    heapq.heappush(hq,(cost,nxt))
        return distance
    
    distance=dijkstra(1)
    m=-int(1e9)
    print(distance)
    for i in range(2,n+1):
        if distance[i]>m:
            m=max(m,distance[i])
            answer=1
        elif distance[i]==m:
            answer+=1
        
    return answer

File: 프로그래머스/3/49189. 가장 먼 노드/가장 먼 노드.java
import java.util.*;

class Solution {
    public int solution(int n, int[][] edge) {
        int answer = 0;
        
        List<List<Integer>> graph = new ArrayList<>();

        for(int i = 0; i <= n; i++){
            graph.add(new ArrayList<>());
        }

        for(int[] e : edge){
            int a = e[0];
            int b = e[1];

            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        int[] distance=new int[n+1];
        Arrays.fill(distance,n+1);
        distance[0]=0;
        Queue<Integer>q=new LinkedList<>();
        q.offer(1);
        distance[1]=0;
        
        while(!q.isEmpty()){
            int now=q.poll();
            for(int next_node:graph.get(now)){
                if(distance[next_node]>distance[now]+1){
                    distance[next_node]=distance[now]+1;
                    q.offer(next_node);                      
                }
            }
        }
        
        Arrays.sort(distance);
        int max_distance=distance[distance.length-1];
        
        for(int dist:distance){
            if(dist==max_distance){
                answer++;
            }
        }
        return answer;
    }
}